Cython/Compiler/ExprNodes.py

```diff
@@ -2342,12 +2342,26 @@ def generate_result_code(self, code):
             # Raise UnboundLocalError for objects and memoryviewslices
             raise_unbound = (
                 (self.cf_maybe_null or self.cf_is_null) and not self.allow_null)
-            null_code = entry.type.check_for_null_code(entry.cname)
 
             memslice_check = entry.type.is_memoryviewslice and self.initialized_check
+            optional_cpp_check = entry.is_cpp_optional and self.initialized_check
 
-            if null_code and raise_unbound and (entry.type.is_pyobject or memslice_check):
-                code.put_error_if_unbound(self.pos, entry, self.in_nogil_context)
+            if optional_cpp_check:
+                unbound_check_code = entry.type.cpp_optional_check_for_null_code(entry.cname)
+            else:
+                unbound_check_code = entry.type.check_for_null_code(entry.cname)
+
+            if unbound_check_code and raise_unbound and (entry.type.is_pyobject or memslice_check or optional_cpp_check):
+                code.put_error_if_unbound(self.pos, entry, self.in_nogil_context, unbound_check_code=unbound_check_code)
+
+        elif entry.is_cglobal and entry.is_cpp_optional and self.initialized_check:
+            null_code = entry.type.cpp_optional_check_for_null_code(entry.cname)
+            code.putln(
+                'if (unlikely(!%s)) {'
+                    'PyErr_SetString(PyExc_NameError,'
+                                    '"C++ global \'%s\' is not initialized");'
+                    '%s'
+                '}' % (null_code, self.name, code.error_goto(self.pos)))
```

Comment:
Could you merge this into `code.put_error_if_unbound()` ? As well as the attribute error raising further down?

It already does some similar checks and then leaves the error raising to a utility function. I think it's good to have a single place for all of these errors. Better than handling some there and some in-place.

(Note that I just moved some of these utility functions into utility code files in the master branch as a cleanup. You can build on that.)